stl/inc/algorithm

```diff
@@ -4639,6 +4639,120 @@ _SampleIt sample(_PopIt _First, _PopIt _Last, _SampleIt _Dest, _Diff _Count,
+                return _Result;
+            }
+
+            _Rng_from_urng<iter_difference_t<_It>, remove_reference_t<_Urng>> _RngFunc(_Func);
+            if constexpr (forward_iterator<_It>) {
+                auto _UResult = _Sample_selection_unchecked(_Get_unwrapped(_STD move(_First)),
+                    _Get_unwrapped(_STD move(_Last)), _Get_unwrapped(_STD move(_Result)), _Count, _RngFunc);
+                _Seek_wrapped(_Result, _STD move(_UResult));
+            } else {
+                auto _UResult = _Sample_reservoir_unchecked(_Get_unwrapped(_STD move(_First)),
+                    _Get_unwrapped(_STD move(_Last)), _Get_unwrapped(_STD move(_Result)), _Count, _RngFunc);
+                _Seek_wrapped(_Result, _STD move(_UResult));
+            }
+            return _Result;
+        }
+
+        template <input_iterator _Rng, weakly_incrementable _Out, class _Urng>
+            requires (forward_range<_Rng> || random_access_iterator<_Out>)
+                  && indirectly_copyable<iterator_t<_Rng>, _Out>
+                  && uniform_random_bit_generator<remove_reference<_Urng>>
+        constexpr _Out operator()(_Rng&& _Range, _Out _Result, range_difference_t<_Rng> _Count, _Urng&& _Func) const {
+            if (_Count <= 0) {
+                return _Result;
+            }
+
+            _Rng_from_urng<iter_difference_t<_It>, remove_reference_t<_Urng>> _RngFunc(_Func);
+            if constexpr (forward_iterator<_It>) {
+                auto _UResult = _Sample_selection_unchecked(
+                    _Ubegin(_Range), _Uend(_Range), _Get_unwrapped(_STD move(_Result)), _Count, _RngFunc);
+                _Seek_wrapped(_Result, _STD move(_UResult));
+            } else {
+                auto _UResult = _Sample_reservoir_unchecked(
+                    _Ubegin(_Range), _Uend(_Range), _Get_unwrapped(_STD move(_Result)), _Count, _RngFunc);
+                _Seek_wrapped(_Result, _STD move(_UResult));
+            }
+            return _Result;
+        }
+        // clang-format on
+    private:
+        template <class _It, class _Se, class _Out, class _Urng>
+        _NODISCARD static constexpr _Out _Sample_selection_unchecked(
+            _It _First, const _Se _Last, _Out _Result, iter_difference_t<_It> _Count, _Urng& _Func) {
+            // randomly select _Count elements from [_First, _Last) into _Result
+            _STL_INTERNAL_STATIC_ASSERT(forward_iterator<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+            _STL_INTERNAL_STATIC_ASSERT(weakly_incrementable<_Out>);
+            _STL_INTERNAL_STATIC_ASSERT(uniform_random_bit_generator<remove_reference<_Urng>>);
+
+            const auto _PopSize = _STD distance(_First, _Last);
+            if (_Count > _PopSize) {
+                _Count = _PopSize;
+            }
+
+            for (; _Count > 0 && _First != _Last; ++_First, (void) --_PopSize) {
+                if (static_cast<iter_difference_t<_It>>(_RngFunc(_PopSize)) < _Count) {

```

Comment:
```suggestion
                if (_RngFunc(_PopSize) < _Count) {
```

`_RngFunc(_PopSize)` is already of type `iter_difference_t<_It>`. The same below.